alica_engine/src/engine/scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,107 @@
+#include "engine/scheduler/Scheduler.h"
+#include "engine/AlicaEngine.h"
+
+#include <algorithm>
+
+namespace alica
+{
+
+namespace scheduler
+{
+
+Scheduler::Scheduler(int numberOfThreads, const alica::AlicaEngine* ae) : _ae(ae), _running(true)
+{
+    for (int i = 0; i < numberOfThreads; i++) {
+        _workers.push_back(new std::thread(&Scheduler::workerFunction, this));
+    }
+}
+
+Scheduler::~Scheduler()
+{
+    _running = false;
+    _jobQueue.clear();
+
+    for (auto worker : _workers) {
+        _workerCV.notify_all();
+        worker->join();
+        delete worker;
+    }
+}
+
+void Scheduler::schedule(std::shared_ptr<Job> job)
+{
+    if (job->scheduledTime.inNanoseconds() == 0) {
+        job->scheduledTime = _ae->getAlicaClock().now();
+    }
+
+    if (job->inExecution || _jobQueue.isScheduled(job)) {
```

Comment:
You should take the mutex here instead of having it in the queue. As it stands you are acquiring/releasing the jobqueue mutex several times in this method, which creates synchronization overhead.